Gracefully Handle a Repository Failing to Clone

mcdonnnj opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

This PR does two things:

1. Update the search used to find appropriate repositories to attempt to make it more deterministic.
2. Make sure that a failure of a `git clone` is handled gracefully.
<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##

There has been an occasional, but recurrent, issue where there are times that a repository is processed twice. This shows up in Actions runs such as https://github.com/cisagov/action-lineage/runs/3279770102?check_suite_focus=true as the error `CRITICAL fatal: destination path 'full_repo_name' already exists and is not an empty directory.`. After looking into the issue I noticed that the repository it failed on was on the page boundary with the default result size (30). This led me to investigate pagination as a possible culprit. I was unable to find anything definitive, but I applied the one change suggested in https://github.com/PyGithub/PyGithub/issues/1748 and wrapped the call to `git clone` in a `try`/`except` to prevent any cloning issue from causing the entire process to fail.
<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##

<!-- How did you test your changes? How could someone else test this PR? -->
Automated testing passes and I checked that the call to `Github.search_repositories()` is successful in the Python command line.
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

## ✅ Checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - _eschew scope creep!_
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All new and existing tests pass.

mcdonnnj committed:
Attempt to make search results more deterministic

I observed an issue where there were duplicates of repositories in the results
of the `get_repo_list()` function. This appears to be caused by an issue with
pagination of the GitHub API on page boundaries. I have added a sort preference
in the hopes of making the search results more deterministic to hopefully
offset this problem.

mcdonnnj committed:
Prevent clone failures from hosing everything

This goes hand-in-hand with the previous commit to hopefully fully resolve the
problem. The failure of a `git clone` call should not result in the entire run
failing, even if it is for a reason other than a duplicate repository.

jsf9k commented on src/lineage/entrypoint.py:
```diff
@@ -91,7 +91,12 @@ def get_repo_list(
     """Generate a list of repositories based on the query."""
     logging.info(f"Querying for repositories: {repo_query}")
-    matching_repos = g.search_repositories(query=repo_query)
+    # Sometimes there is an issue with the pagination of results from the
+    # call to `Github.search_repositories()` that causes repositories on page
+    # breaks to be duplicated when iterating over results. The only issue I could
+    # find that referenced this problem is
+    # https://github.com/PyGithub/PyGithub/issues/1748
+    matching_repos = g.search_repositories(query=repo_query, sort="updated")
```
```suggestion
    # Sometimes there is an issue with the pagination of results from the
    # call to `Github.search_repositories()` that causes repositories on page
    # breaks to be duplicated when iterating over results. The only issue I could
    # find that referenced this problem is
    # https://github.com/PyGithub/PyGithub/issues/1748,
    # which suggests that adding the sort="updated" parameter to the
    # search_repositories() call helps.  Hence I am adding it here.
    #
    # See also issue #33 in this repository:
    # https://github.com/cisagov/action-lineage/issues/33
    matching_repos = g.search_repositories(query=repo_query, sort="updated")
```

mcdonnnj replied:
Excellent feedback! Please see 3d7c955 as I had to make some edits to your suggestion.

jsf9k commented on src/lineage/entrypoint.py:
```diff
@@ -314,7 +319,14 @@ def main() -> None:
         logging.info(f"Lineage configuration found for {repo.full_name}")
         logging.info(f"Cloning repository: {repo.clone_url}")
-        run([GIT, "clone", repo.clone_url, repo.full_name])
+        # The failure of a repository to clone should not cause the entire
+        # run to fail.
```
```suggestion
        # The failure of a repository to clone should not cause the entire
        # run to fail.  See related comment above in the body of the 
        # get_repo_list() function.
        #
        # See also issue #33 in this repository:
        # https://github.com/cisagov/action-lineage/issues/33
```

mcdonnnj replied:
Good call. Please see 3d7c955 as I had to make some edits to your suggestion.

jsf9k reviewed (approved):
I suggested fleshing out a few comments to remind future us why we did these things.

mcdonnnj committed:
Expand explanatory comments after feedback

Expand the existing comments to explain what was done in more detail and to add
references to the pull request for more information.

Co-authored-by: Shane Frasier <jeremy.frasier@trio.dhs.gov>

dav3r reviewed (approved):
Nice changes and comments to go along with them. 👍
